src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,208 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "AsyncTransferFacilitator.h"
+
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace bdx {
+
+AsyncTransferFacilitator::~AsyncTransferFacilitator() {}
+
+bdx::StatusCode AsyncTransferFacilitator::GetBdxStatusCodeFromChipError(CHIP_ERROR err)
+{
+    if (err == CHIP_ERROR_INCORRECT_STATE)
+    {
+        return bdx::StatusCode::kUnexpectedMessage;
+    }
+    if (err == CHIP_ERROR_INVALID_ARGUMENT)
+    {
+        return bdx::StatusCode::kBadMessageContents;
+    }
+    return bdx::StatusCode::kUnknown;
+}
+
+CHIP_ERROR AsyncTransferFacilitator::Init(System::Layer * layer, Messaging::ExchangeContext * exchangeCtx,
+                                          System::Clock::Timeout timeout)
+{
+    VerifyOrReturnError(layer != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(exchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(!mExchange, CHIP_ERROR_INCORRECT_STATE);
+
+    mSystemLayer = layer;
+    mExchange.Grab(exchangeCtx);
+    mTimeout                = timeout;
+    mProcessingOutputEvents = false;
+    mDestroySelf            = false;
+    return CHIP_NO_ERROR;
+}
+
+/**
+ * Calls the GetNextAction on the TransferSession to get the next output events until it receives
+ * TransferSession::OutputEventType::kNone If the output event is of type TransferSession::OutputEventType::kMsgToSend, it sends the
+ * message over the exchange context, otherwise it calls the HandleTransferSessionOutput method implemented by the subclass to
+ * handle the BDX message.
+ */
+void AsyncTransferFacilitator::ProcessOutputEvents()
+{
+    if (mProcessingOutputEvents)
+    {
+        ChipLogDetail(BDX, "ProcessOutputEvents: Still getting and processing output events from a previous call. Return.");
+        return;
+    }
+
+    mProcessingOutputEvents = true;
+
+    // Get the next output event and handle it based on the type of event.
+    // If its of type kMsgToSend send it over the exchange, otherwise call the HandleTransferSessionOutput
+    // virtual method that must be implemeted by the subclass of this class to handle the BDX message.
+    TransferSession::OutputEvent outEvent;
+
+    mTransfer.GetNextAction(outEvent);
+    while (outEvent.EventType != TransferSession::OutputEventType::kNone)
+    {
+        if (outEvent.EventType == TransferSession::OutputEventType::kMsgToSend)
+        {
+            CHIP_ERROR err = SendMessage(outEvent.msgTypeData, outEvent.MsgData);
+
+            // If we failed to send the message across the exchange, there is no way to let the other side know there was an
+            // error so call DestroySelf so the exchange can be released and the other side will be notified the exchange is
+            // closing and will clean up.
```

Comment:
I don't get this comment.  I _think_ what this is trying to say is:
```suggestion
            // If we failed to send the message across the exchange, just abort the transfer.
            // We have no way to notify our peer we are doing that (we can't send them a
            // message!) but eventually they will time out.
```